### ISSUE TITLE
Project status + ideas

### ISSUE THREAD

**Issue by doonte** [athanorlabs/atomic-swap#512]
@noot @dimalinux What is the current status of this project?

Fees are high on mainnet. I could see this being used more on L2s like Optimism or Arbitrum or Base.

Ideas for revival, maybe in a separate project:
* Rewrite so that everything is done from a front-end interface and have the user make transactions from their native wallet instead of a monero daemon loaded with funds (more prone to missed windows if the user does not pay attention, but be clear with this to users)
* Use i2p for finding peers, no IP leaks
* Native eth L2 support high priority

**Comment by dimalinux** [comment]
@doonte I'd say the project right now is on maintenance mode. If something breaks, I'll fix it. Moving the project away from Ganache, which is no longer supported, is still on my TODO list, since we'll need the switch for continued testing. I would have already added support for L2s, but my concern is we that don't want to put too much effort into a technology that probably won't take off even with a 1st class GUI. If you have reason to believe otherwise, chime in. My thought is that systems that use liquidity pools will always provide a better user experience and better liquidity. I've started learning Rust so I can get my hands dirty with Serai or tooling around Serai.

**Comment by doonte** [comment]
Thank you for your response. I agree that liquidity pools will probably always be a better user experience. I see that Serai suggests it will use its own coin (SRI) for fees - it's a little disappointing to me that it doesn't make use of a more established coin or the coins provided as liquidity instead.

**Comment by kayabaNerve** [comment]
> I see that Serai suggests it will use its own coin (SRI) for fees 

That's not the use of SRI.

SRI is used as one side of every pool for efficient routing (we don't fragment ETH liquidity as ETH-BTC and ETH-XMR). We could use XMR as one side, yet then we don't have a value oracle for BTC/ETH/XMR in terms of SRI. We need a value oracle in terms of SRI to ensure economic security (the SRI stake necessary to misbehave with revenue has greater value than the revenue which would be gained). If we used XMR as stake, then using XMR for every pool would establish a value oracle in XMR terms and that'd be fine.

The issue is we can't use XMR as stake. It'd only be slashable assuming an honest multisignature, yet we need economic security to hold upon a dishonest multisig. We need a coin:
1) We can program the conditions on for partial failures
2) Is subject to market forces upon complete system failure

For more context on the latter, please read https://github.com/kayabaNerve/notes/blob/main/EconomicSecurity.pdf.

**Comment by StrawberryChocolateFudge** [comment]
I am interested in using the project to compile composable WASM modules to use as a library later, for building a browser only atomic swap.

E.G: I took out the dleq and created a standalone wasm module so I can call it from JS. https://github.com/StrawberryChocolateFudge/go-dleq-wasm

Need to do the same for the monero crypto used for summing the keys, then I think it should be possible to build a browser native swap using the monero-ts Local Full Wallet.

I'm also thinking to lift out the solidity contracts from the project and create a separate importable module with Hardhat.

I am however not sure if a browser based swap should use the same libp2p backend to be compatible 100%, so I'm just thinking to make a few critical building blocks from the existing code and see if I can use it for something.

**Comment by dimalinux** [comment]
@StrawberryChocolateFudge It's an interesting idea, but I would spend some time thinking about two protocol-level issues with all the Monero atomic swap protocols and see if you have any ideas on how to solve them:

1. In order for the swap protocol to succeed, the XMR provider should have skin in the game. If an XMR provider starts a swap and doesn't complete it, it can't be free. Otherwise the protocol will get DOS'ed as soon as it gets popular.  The ETH side has skin in the game both in terms of transaction fees and the extended period during which their ETH is locked on chain until the swap fails.
2. Bidirectional: A design where either side can be the maker.

Until those issues have some good ideas on how they can be solved, I'm crossing my fingers for non-atomic-swap solutions like Serai DEX to come to fruition.

**Comment by StrawberryChocolateFudge** [comment]
@dimalinux 

I see, so for the XMR provider to have skin in the game they would need to make a security deposit? 

What about introducing a protocol fee?

Maybe a small amount of XMR that equals the Eth gas fee could be deposited into a multisig address with published secret view key.

Kind of like:

XMR provider makes a small payment that equals an Eth gas fee for a swap, to make fees on both sides equal.

Eth provider verifies the protocol fee was paid.

The fee payment proof could be hashed and stored in the smart contract and then it's nullified, so it can't be reused.

But if the swap is not completed because the ETH side didn't do it's thing, then the proof of fee payment is valid for the next swap and the XMR side doesn't need to pay again.

This way it costs the same for both sides.  The only way to DOS this would be if the XMR provider is talking simultaneously to many Eth providers, so Eth providers would need to gossip to each other to verify this is not happening.

**Comment by dimalinux** [comment]
> I see, so for the XMR provider to have skin in the game they would need to make a security deposit?

Yes, a security deposit would add some skin in the game. Not easy, as Monero doesn't have smart contracts or scripting.

> What about introducing a protocol fee?

We'd need a full plan on how it gets collected, incentives, etc.

> Maybe a small amount of XMR that equals the Eth gas fee could be deposited into a multisig address with published secret view key.

It might need to be more than the ETH gas fee to properly align incentives. Imagine that we implement the protocol on an EVM blockchain with super low gas fees. In that case, the ETH side still has the cost of locking their funds in a swap contract for an hour. During that time, the ETH side can't carry out a swap with a legit counter-party. That said, if the XMR side has enough skin in the game, you now need a mechanism to ensure that a malicious ETH side can't slowly drain out the funds from all automated XMR makers without paying some kind of consequence.

**Comment by StrawberryChocolateFudge** [comment]
I see, thanks. 

I think with a protocol fee, the ETH side can't drain funds If the ETH provider has to nullify a "proof of payment" when locking funds on-chain. 

1. XMR provider makes a payment of x amount , to fee address with published secret view key to pay the protocol fee

2. XMR provider has proof TXID,TXKEY,ADDRESS and the nullifier for it is HASH(TXID,TXKEY,ADDRESS) 

3. ETH provider verifies the payment was made with check_tx_key TXID TXKEY ADDRESS and by fetching the transaction id using the view only wallet

4. Then ETH provider computes HASH(TXID,TXKEY,ADDRESS) and check the SwapCreator if HASH is nullified, by calling a function isNullified(hash) => bool 

5. if not then calls the newSwap function with an extra bytes32 nullifier argument, which nullifies fee payment by saving a bool into a mapping. mapping(bytes32 => bool) isNullified
The newSwap function should revert if isNullified(nullifer) == true 

If the ETH provider does not lock funds and attempts to drain the XMR provider via fee, the attack does not work

Because as long as the fee payment is not nullified, the XMR provider doesn't need to pay the fee again.

If the XMR provider wants to DOS the ETH provider, it's not free and needs to pay a fee for each attack.
